File: 1011. Capacity To Ship Packages Within D Days/Solution.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int left = Integer.MIN_VALUE;
        int right = 0;

        for(int i=0;i<weights.length;i++){
            left = Math.max(left,weights[i]);
            right+=weights[i];
        }

        int ans = right;

        while(left<=right){
            int mid = left + (right-left)/2;

            if(isPossible(weights,days,mid)){
                ans = mid;
                right = mid-1;
            }
            else{
                left = mid+1;
            }
        }
        return ans;
    }
    private boolean isPossible(int[] weights,int days,int c){
        int total = 0;
        int daysTaken = 1;
        for(int i : weights){
            if((total+i)>c){
                daysTaken+=1;
                total = i;

                if(daysTaken>days)return false;
            }
            else{
                total += i;
            }  
        }
        return days>=daysTaken;
    }
}

File: 1011. Capacity To Ship Packages Within D Days/Solution.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        """
        :type weights: List[int]
        :type days: int
        :rtype: int
        """
        left = max(weights)
        right = sum(weights)
        ans = right

        while left<=right:
            mid = left + (right-left)//2

            if self.isPossible(weights,days,mid):
                ans = mid
                right = mid-1
            else:
                left = mid+1
        return ans

    def isPossible(self,weights,days,c):
        total = 0
        daysTaken = 1

        for i in weights:
            if total+i>c:
                daysTaken+=1
                total = i

                if daysTaken>days:
                    return False
            else:
                total+=i
        return daysTaken<=days

